code/Control/UserDefinedFormController.php

```diff
@@ -511,6 +533,250 @@ public function finished()
+
+        /** @var Folder $formSubmissionsFolder */
+        $formSubmissionsFolder = Folder::find($folder);
+        if ($formSubmissionsFolder) {
+            $groups = $formSubmissionsFolder->ViewerGroups()->column('Title');
+        } else {
+            $groups = [Group::get_one(Group::class, ['Code' => 'administrators'])->Title];
+        }
+        $textField->setDescription(
+            '<span class="font-icon-eye pr-1 align-middle" aria-hidden="true"></span>' .
+            _t(
+                __CLASS__ . '.RESTRICTEDACCESS',
+                "Restricted access, uploads will be visible to the following groups"
+            ) .
+            ': ' . implode(', ', $groups)
+        );
+        $textField->addExtraClass('pt-2');
+        $textField->setSchemaData([
+            'data' => [
+                'prefix' => static::config()->get('form_submissions_folder') . '/',
+            ],
+        ]);
+
+        $textField->setAttribute('data-prefix', static::config()->get('form_submissions_folder'));
+        $textField->setAttribute('prefix', static::config()->get('form_submissions_folder'));
+
+        return $textField;
+    }
+
+    /**
+     * This returns a Confirm Folder form used to verify the upload folder for EditableFileFields
+     * @return ViewableData
+     */
+    public function confirmfolderform()
+    {
+        $request = $this->getRequest();
+        $id = $request->requestVar('ID');
+        $userFormID = $request->requestVar('UserFormID');
+        $userForm = UserDefinedForm::get()->byID($userFormID);
+        if (!$userForm) {
+            $userForm = Versioned::get_by_stage(UserDefinedForm::class, Versioned::DRAFT)->byID($userFormID);
+        }
+
+        if (!$userForm->canView()) {
+            throw new PermissionFailureException();
+        }
+
+        $editableFormField = EditableFormField::get()->byID($id);
+        if (!$editableFormField) {
+            $editableFormField = Versioned::get_by_stage(EditableFormField::class, Versioned::DRAFT)->byID($id);
+        }
+
+        $folderId = 0;
+        if (!$editableFormField) {
+            $editableFormField = new EditableFileField();
+        } elseif ($editableFormField instanceof EditableFileField) {
+            $folderId = $editableFormField->Folder->ID;
+        }
+        /** @var Folder $folder */
+        $folder = Folder::get()->byID($folderId);
+
+
+        $fields = new FieldList();
+
+
+        $label = new LiteralField('Label', _t(__CLASS__.'.CONFIRM_FOLDER_LABEL', 'Files that your users upload should be stored carefully to reduce the risk of exposing sensitive data. Ensure the folder you select can only be viewed by appropriate parties. Folder permissions can be managed within the Files area.'));
+        $label->addExtraClass(' mb-3');
+        $fields->push($label);
+
+
+        $fields->push(static::getRestrictedAccessField($this->config()->get('form_submissions_folder'), $userForm->Title));
```

Comment:
It should be static if it can be static.

Reply:
Because of the way we use Injector and singletons, we've tended to prefer using instance methods over class methods, even in cases where other frameworks would use `static`.

It's not a big deal, but it's something to keep in mind